### PR TITLE
Add --bug-report flag

### PR DESCRIPTION
**Pull request by tvolkert**
This adds support for a `--bug-report` flag, which is a recording
that:
  - includes the arguments that were passed to the command runner
  - is zipped up for easy attachment in Guthub issues

### CODE REVIEW

**Comment by tvolkert** [comment]
@johnmccutchan

File: packages/flutter_tools/lib/src/base/file_system.dart [hunk 1]
```diff
@@ -34,10 +34,14 @@ FileSystem get fs => context == null ? _kLocalFs : context[FileSystem];
 /// It is permissible for [location] to represent an existing non-empty
 /// directory as long as there is no collision with the `"file"` subdirectory.
 void enableRecordingFileSystem(String location) {
+  FileSystem originalFileSystem = fs;
   Directory dir = getRecordingSink(location, _kRecordingType);
   RecordingFileSystem fileSystem = new RecordingFileSystem(
       delegate: _kLocalFs, destination: dir);
-  addShutdownHook(() => fileSystem.recording.flush());
+  addShutdownHook(() async {
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/base/file_system.dart]: general question- how do we handle ordering issues with these shutdown hooks?

File: packages/flutter_tools/lib/src/base/process.dart [hunk 2]
```diff
@@ -11,26 +11,70 @@ import 'io.dart';
...
 
 typedef String StringConverter(String string);
+
+/// A function that will be run before the VM exits.
 typedef Future<dynamic> ShutdownHook();
 
 // TODO(ianh): We have way too many ways to run subprocesses in this project.
 
-List<ShutdownHook> _shutdownHooks = <ShutdownHook>[];
+/// The stage in which a [ShutdownHook] will be run. All shutdown hooks within
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/base/process.dart]: Oh wait, I see you've anticipated my above question!

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 3]
```diff
@@ -97,14 +102,14 @@ class FlutterCommandRunner extends CommandRunner<Null> {
             'Use this to select a specific version of the engine if you have built multiple engine targets.\n'
             'This path is relative to --local-engine-src-path/out.');
     argParser.addOption('record-to',
-        hide: !verboseHelp,
+        hide: true,
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: These will never be presented to the user?

File: packages/flutter_tools/lib/src/base/process.dart [hunk 4]
```diff
@@ -11,26 +11,70 @@ import 'io.dart';
...
 Future<Null> runShutdownHooks() async {
-  List<ShutdownHook> hooks = new List<ShutdownHook>.from(_shutdownHooks);
-  _shutdownHooks.clear();
   _shutdownHooksRunning = true;
   try {
-    List<Future<dynamic>> futures = <Future<dynamic>>[];
-    for (ShutdownHook shutdownHook in hooks)
-      futures.add(shutdownHook());
-    await Future.wait<dynamic>(futures);
+    for (ShutdownStage stage in _shutdownHooks.keys.toList()..sort()) {
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/base/process.dart]: Might be nice to have a simple unit test that verifies these stages happen in a specific order. More as a regression test than anything else.

**Review by johnmccutchan — COMMENTED** [review]
LGTM w/c

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 3 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Correct.  A script to create golden files will use it, and `--bug-report` wraps it, but users won't be directly exposed to it.

File: packages/flutter_tools/lib/src/base/process.dart (hunk 4 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/process.dart]: Done

**Pull request merged**